Add Product Query Support for Atomic SKU Block

danielwrobert opened the pull request:
Adds Product Query support for the atomic SKU block.

On the client side, when the SKU Block is used within the Product Query block, the markup will be rendered on the server side - no javascript related to the SKU block will be loaded.

<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->

<!-- Reference any related issues or PRs here -->

Fixes #7332 

<!-- Don't forget to update the title with something descriptive. -->
<!-- If you can, add the appropriate labels -->


### Testing

1. Checkout this branch and run `npm start` (need to run the `start` command to catch the experimental flags).
1. Be sure that you have the latest version of the Gutenberg feature plugin installed and activated.
1. Add the `woocommerce/product-sku` block to the Product Query `INNER_BLOCKS_TEMPLATE` in `assets/js/blocks/product-query/constants.ts` under line 54 (**only for testing, remove afterwards**).
1. Create a post/page and add the Product Query block.
1. Save it.
1. Open the frontend side and confirm that:
    - No JS related to the Product SKU block is loaded (search in your network tab for `product-sku-frontend.js`).
    - The UI of the block is identical to on the editor side.
    - That the settings are reflected correctly (eg: custom style).
1. Back in the editor confirm that the Product SKU block can be added as a child of the Product query block via the inserter (inside **Product Query > Post Template** use **Insert Before/After** one of the existing children blocks).
1. For comparison, switch to trunk and add the All Products block to confirm that the aforementioned `product-sku-frontend.js` _is_ loaded.

* [x] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [ ] WooCommerce Core
* [ ] Feature plugin
* [x] Experimental

danielwrobert committed:
Add attributes, settings, and editor PQ settings.

- Adds isDescendentOfQueryLoop attribute and sets up usage in editor.
- Connects Context (via useContext) in editor.
- Sets up necessary hierarchy in block index file settings.

danielwrobert committed:
Update parent inner blocks config.

Reassign parent array to ancestor array which allows for blocks to be
included with more flexibility - i.e., added within groups that are
children of the ancestor block.

danielwrobert committed:
Add dynamic render function for PQ support.

imanish003 commented:
> The UI of the block is identical to on the editor side.

For me, UI looks different in editor vs frontend. 
| Editor | Frontend |
| - | -  |
| ![image](https://user-images.githubusercontent.com/16707866/200787557-d9c6b96d-7439-4cfd-8916-d2f7943a2677.png) | ![image](https://user-images.githubusercontent.com/16707866/200787632-adea4c34-8436-42f6-9c02-3f6c50340fe0.png) |

imanish003 reviewed (changes requested):
> The UI of the block is identical to on the editor side.

For me, UI looks different in editor vs frontend. 
| Editor | Frontend |
| - | -  |
| ![image](https://user-images.githubusercontent.com/16707866/200787557-d9c6b96d-7439-4cfd-8916-d2f7943a2677.png) | ![image](https://user-images.githubusercontent.com/16707866/200787632-adea4c34-8436-42f6-9c02-3f6c50340fe0.png) |

----

Also, in this [PR for the Rating block](https://github.com/woocommerce/woocommerce-blocks/pull/7352), you converted js to ts. I am not sure if that is something we wanna cover in this PR too.

danielwrobert committed:
Update to use correct classnames on PHP side.

SSR markup was accidentally using rating classname instead of sku so it
was not applying the proper text transform seen in the editor.

danielwrobert commented:
@imanish003 nice catch, thank you! Fixed in d6952b0c4.

The TS conversion took place separately but I have rebased this PR to include those changes from `trunk`.

github-actions[bot] committed:
bot: update checkstyle.xml

danielwrobert committed:
Rename BlockAttributes type to Attributes

This keeps things consistent with the type naming on the other blocks.

danielwrobert committed:
Remove redundant spread of sharedConfig object.

We don't need to spread the `sharedConfig` object into the `blockConfig`
object when defining `blockConfig` since we spread/merge these two
objects when registering the block via `registerExperimentalBlockType`.

danielwrobert commented on assets/js/atomic/blocks/product-elements/sku/index.ts:
```diff
@@ -17,11 +17,16 @@ import {
 } from './constants';
 
 const blockConfig: BlockConfiguration = {
-	...sharedConfig,
```
Just want to confirm we don't need this here, as we're spreading/merging the same config object in the `registerExperimentalBlockType` function below.

It seems redundant to spread it here in this definition, unless I'm missing something.

imanish003 reviewed (approved):
Verified, Now UI on Editor and Frontend looks the same 🚀

github-actions[bot] committed:
bot: update checkstyle.xml

danielwrobert committed:
Empty commit to trigger checks.
